# Profitable App Profiles for *App Store* and *Google Play*
---

Our aim in this project is to find mobile app profiles that are profitable for the *App Store* and *Google Play* markets. We're working as data analysts for a private company that builds Android and iOS mobile apps, and our job is to enable our team of developers to make data-driven decisions with respect to the kind of apps they build.

At our [company](https://app.dataquest.io/), we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that our revenue for any given app is mostly influenced by the number of users that use our app. Therefore our main criteria are:
1. free apps
2. english apps

Our goal for this project is to analyze data to help our developers understand what kinds of apps are likely to attract more users.


## Candidate Data Sets
---

As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.(Source: [Statista](https://www.statista.com/statistics/276623/number-of-apps-available-in-leading-app-stores/))

Collecting data for over 4 million apps requires a significant amount of time and money. So we'll try to analyze a sample of the data instead. These two data sets seem suitable for our goals:

1. [A data set](https://www.kaggle.com/lava18/google-play-store-apps) containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018. You can download the data set directly from [this link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv).
2. [A data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017. You can download the data set directly from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv).

### Column description for `googleplaystore.csv` data set is given below:

|Column index| Column name  | Column Description                                                |
| :--------: | :----------: | :---------------------------------------------------------------: |
|00          |App           |Application name                                                   |
|01          |Category      |Category the app belongs to                                        |
|02          |Rating        |Overall user rating of the app                                     |
|03          |Reviews       |Number of user reviews for the app                                 |
|04          |Size          |Size of the app                                                    |
|05          |Installs      |Number of user downloads/installs for the app                      |
|06          |Type          |Paid or Free                                                       |
|07          |Price         |Price of the app                                                   |
|08          |Content Rating|Age group the app is targeted at                                   |
|09          |Genres        |An app can belong to multiple genres (apart from its main category)|
|11          |Last Updated  |Date when the app was last updated on Play Store                   |
|12          |Current Ver   |Current version of the app available on Play Store                 |
|13          |Android Ver   |Min required Android version (as when scraped)                     |


### Column description for `AppleStore.csv` data set is given below:

|Column index| Column name      | Column Description                             |
| :--------: | :--------------: | :--------------------------------------------: |
| 00         | "id"             | App ID                                         |
| 01         | "track_name"     | App Name                                       |
| 02         |"size_bytes"      | Size (in Bytes)                                |
| 03         |"currency"        | Currency Type                                  |
| 04         |"price"           | Price amount                                   |
| 05         |"ratingcounttot"  |User Rating counts (for all version)            |
| 06         |"ratingcountver"  |User Rating counts (for current version)        |
| 07         |"user_rating"     | Average User Rating value (for all version)    |
| 08         |"userratingver"   | Average User Rating value (for current version)|
| 09         |"ver"             | Latest version code                            |
| 10         |"cont_rating"     | Content Rating                                 |
| 11         |"prime_genre"     | Primary Genre                                  |
| 12         |"sup_devices.num" | Number of supporting devices                   |
| 13         |"ipadSc_urls.num" | Number of screenshots showed for display       |
| 14         |"lang.num"        | Number of supported languages                  |
| 15         |"vpp_lic"         | Vpp Device Based Licensing Enabled             |

## Opening the Data

-----------

At first we are going to cache/load the data from the datasets into memory for faster processing. We will load the data into python list.

To open and create a list containing the data set, we'll create a function `csv_to_list()` which consists of the following paramters:

1. str `file_path` is the file path to our dataset 
2. boolean `seperate_header` indicates if we want to seperate the header, default: True
3. int `header_row` points to the row index of the header, default: 0
4. str `encoding` provides additional encoding information, default: 'utf8' 

We have set parameters 2,3 and 4 with default arguments according to our needs. Encoding is set to utf8 due to UnicodeDecodeError.

The function `csv_to_list()` opens the dataset, reads it and stores the whole dataset in a single list called `dataset`. Depending on the `seperate_header` argument, the function either returns `dataset` without seperating the header row. Or it returns a list of the header row (index 0) and a list of the dataset (index 1). We choose the later by default.

In [1]:
def csv_to_list(file_path, seperate_header=True ,header_row=0, encoding='utf8'):
    from csv import reader
    open_file = open(file_path, encoding=encoding)
    read_file = reader(open_file)
    dataset = list(read_file)
    
    if seperate_header:
        header_list = dataset[header_row]
        data_list = dataset[header_row+1:]
        return [header_list, data_list]
    else:
        return dataset

By running the `csv_to_list()` function for both the datasets, we get lists that have the header and the data seperated at index 0 and 1 respectively.

In [2]:
[ios_header, ios_list] = csv_to_list('AppleStore.csv')
[android_header, android_list] = csv_to_list('googleplaystore.csv')

## Exploring the Data
---

Now we shall explore the datasets to understand it's contents and strategize how we can clean and analyze the data. The column descriptions (header description) in the introduction is very helpful here to understand the output.

To explore these list of data, we'll first write a function `explore_data()` which will print the some rows indicated by the following paramters:

1. list `dataset` is the data set represented as a list which we want to explore
2. int `start` is the row index from where we'll start to print the informations of
3. int `end` is the row index upto (not including) which we'll print the informations of
4. boolean `rows_and_cloumns` indicates whether we want to print the number of rows and columns of the `dataset` list; default: False

We have set parameter 4 as False to not print the number of rows and columns by default.

The function `explore_data()` prints the informations of a list of `dataset` from the `start` index upto (not including) the `end` index. The function also has a trigger to show the number of rows and columns for any `dataset` which is turned off by default.

In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line between rows
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

---
We run the `explore_data()` function on the *App Store* dataset and check out the first two rows and the number of rows and columns the dataset consists.

In [4]:
print(ios_header)
print('\n')
explore_data(ios_list, 0, 2, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7197
Number of columns: 16


We see that the data set has 7197 iOS apps (each row represents data about an app) and 16 columns. The columns that seem interesting are:<br/>
`'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver',` and `'prime_genre'`

---
Running the `explore_data()` function on the *Google Play* data set and check out the first two rows and the number of rows and columns the dataset consists.

In [5]:
print(android_header)
print('\n')
explore_data(android_list, 0, 2, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


We see that the data set has 10841 Android apps (each row represents data about an app) and 13 columns. The columns that might be useful for the purpose of our analysis are: <br></br>
`'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price',` and `'Genres'` 

## Checking For Wrong Data
---
The Google Play data set has a dedicated [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), where we can see that [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472. So we need to find out the rows that have missing data (column missing) in and exclude them from our analysis. We are basically cleaning the data.

Let's create a function `missing_data()` which will return a list of row indices that have values missing using the parameters:

1. list `header_row` is only the header row (containg column names) represented as a list
2. list `data_set` is the remaining data excluding the header row represented as a list

The paramaters do not have any default arguments.


The `missing_data()` function will compare the length of the `header_row` with the length of each rows in the `data_set` to determine if any row is missing any values. The function then returns a list of the row indices which are missing data.

In [6]:
def missing_data(header_row, data_set):
    len_header =len(header_row)
    missing_indices = []
    
    index = 0
    for row in data_set:
        if len(row) != len_header:
            missing_indices.append(index)
        index += 1
        
    return missing_indices
            

---
Let's use the `missing_data()` function to genrate a list of row indices which are missing some data in the `android_list`. Let's also cross check to see if the list `ios_list` generated from the *AppleStore.csv* data set is missing any data.

In [7]:
missing_android = missing_data(android_header, android_list)
print('android_list, missing data on indices:', missing_android)

missing_ios = missing_data(ios_header, ios_list)
print('ios_list, missing data on indices:', missing_ios)

android_list, missing data on indices: [10472]
ios_list, missing data on indices: []


According to our `missing_data()` function, the list `ios_list` generated from the *'AppleStore.csv'* data set has no missing data.
But for the *'googleplaystore.csv'* data set, we can see that [this discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlined an error for row 10472, which was also spotted by our function. 


---
Let's print row 10472 and compare it against it's header and another row that is correct.

In [8]:
print(android_list[10472])  # incorrect row
print('\n')
print(android_header)  # header
print('\n')
print(android_list[0])      # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


The row 10472 corresponds to the app 'Life Made WI-Fi Touchscreen Photo Frame', and we can see that the rating is 19. This is clearly off because the maximum rating for a Google Play app is 5. If we look at all the columns of this row we can observe that there has been a column shift. As a consequence, we cannot work with this row so we'll delete it.


## Deleting Wrong Data
---

We want to delete some rows that we do not want to consider on our analysis for various reasons.

We'll make a function `delete_rows()` which will delete rows according to the parameters:

1. list `data_set` contains the data excluding the header and is represented as a list
2. list `delete_indices` contains a list of row indices that are to be deleted

The paramters do not have any default arguments.

The function `delete_rows()` takes a list `data_set` and a list `delete_indices` to delete the listed rows in it.

In [9]:
def delete_rows(data_set, delete_indices):
    deleted = 0
    for index in delete_indices:
        del data_set[index - deleted]
        deleted += 1

---
Before deleting the rows, we first print the rows 10471, 10472 and 10473 which we will use to compare in the upcoming steps.

In [10]:
print(android_list[10471])
print(android_list[10472])
print(android_list[10473])

['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


---
Let's run the function `delete_rows()` to delete the selected rows of `android_list` by providing the list `missing_android` generated by the `missing_data()` function as arguments. We also print the length of `android_list` before and after running the `delete_rows()` function.

In [11]:
print('before deleting:', len(android_list))
delete_rows(android_list, missing_android)
print('after deleting:', len(android_list))

before deleting: 10841
after deleting: 10840


We can see that the function was successful as the number of rows have decreased after running it is just 1.

---

We again print the rows 10471, 10472 and 10473 to see if we get the intended change.

In [12]:
print(android_list[10471])
print(android_list[10472])
print(android_list[10473])

['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']
['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']
['Sat-Fi Voice', 'COMMUNICATION', '3.4', '37', '14M', '1,000+', 'Free', '0', 'Everyone', 'Communication', 'November 21, 2014', '2.2.1.5', '2.2 and up']


Since the row 10472 now represents the data of it's next row (10473), so we can conclude that we ran a successful delete operation.

## Checking For Duplicate Entries
----


If we explore the Google Play data set long enough or look at [this discussions section](https://www.kaggle.com/lava18/google-play-store-apps/discussion) we can notice that some apps have duplicate entries. For instance, Instagram has four entries:

In [13]:
for row in android_list:
    app_name = row[0]
    if app_name == 'Instagram':
        print(row)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


---

Duplicate entries should be removed based on analysis criteria. 

Let's first check to see the total number of duplicate entries using the function `list_duplicates()` having the parameters:

1. list `data_set` is a list which represents a dataset
2. int `key_column` is an index of a column which is supposed to be unique (key attribute)

The function doesn't have any default argument.

The `list_duplicates()` function returns a list of duplicate values based on the supposed unique column index `key_column` of a list `data_set`.

In [14]:
def list_duplicates(data_set, key_column):
    duplicates = []
    uniques = []
    
    for row in data_set:
        key = row[key_column]
        if key in uniques:
            duplicates.append(key)
        else:
            uniques.append(key)
    
    return duplicates

---
We use the `list_duplicates()` function to generate a list of duplicate app names on our data sets `android_list` and `ios_list`. We consider app names to be unique, so we find app names that occurs more than once. We also print the length of the generated list for both cases and some duplicate app names.

In [15]:
duplicate_androidapps = list_duplicates(android_list, 0)
print('Number of duplicate android apps:', len(duplicate_androidapps))
print('\n')
print('Examples of duplicate android apps:', duplicate_androidapps[:15])
print('\n')

duplicate_iosapps = list_duplicates(ios_list, 0)
print('Number of duplicate ios apps:', len(duplicate_iosapps))
print('\n')
print('Examples of duplicate ios apps:', duplicate_iosapps[:15])

Number of duplicate android apps: 1181


Examples of duplicate android apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


Number of duplicate ios apps: 0


Examples of duplicate ios apps: []


In total, there are 1,181 cases where the name of an android app occurs more than once. Fortunately there are no duplicate ios apps.

## Removing Duplicate Entries
---
Based on the information of the 4 Instagram entries we can see that although the app name is the same, Instagram but the number of reviews are different. So we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings and remove the other duplicates.

To do that, we will:

1. Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
2. Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)
---
Let's start by building the dictionary.

In [16]:
reviews_max = {}

for row in android_list:
    app_name = row[0]
    num_reviews = float(row[3])
    if (app_name in reviews_max) and (num_reviews > reviews_max[app_name]):
        reviews_max[app_name] = num_reviews
    elif app_name not in reviews_max:
        reviews_max[app_name] = num_reviews

---
In a previous code cell, we found that there are 1,181 cases where an app occurs more than once, so the length of our dictionary (of unique apps) should be equal to the difference between the length of our data set and 1,181.

In [17]:
print('Expected length:', len(android_list) - len(duplicate_androidapps))
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


Since the two lengths are equal that means the operation was successful and `reviews_max` contains unique app names as the key and the highest number of reviews as the value.

---
Now, let's use the `reviews_max` dictionary to remove the duplicates. For the duplicate cases, we'll only keep the entries with the highest number of reviews. In the code cell below:

1. We start by initializing two empty lists, `android_clean` and `rows_added`.
2. We loop through the `android_list` dataset, and for every iteration:

    a. We isolate the name of the app and the number of reviews.
    
    b. We add the current row to the `android_clean` list, and the app name to the `already_added` list if: the number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary. And the name of the app is not already in the `rows_added` list. 

It is to be noted that, the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries with the same number of reviews). If we just check for the number of reviews, we'll still end up with duplicate entries for some apps.

In [18]:
android_clean = []  # new cleaned data set
rows_added = []  # app names in android_clean

for row in android_list:
    app_name = row[0]
    num_reviews = float(row[3])
    if (num_reviews == reviews_max[app_name]) and (app_name not in rows_added):
        android_clean.append(row)
        rows_added.append(app_name)
    

---
Now let's quickly explore the new data set, and confirm that the number of rows is 9,659.

In [19]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13



We have 9659 rows, just as expected.

## Checking For Non-English Apps
---
In both the data sets, we can notice that the names of some of the apps are not English, which suggests they are not directed toward an English-speaking audience. We're not interested in keeping these kind of apps, so we'll remove them. We can create a function which checks an app name and tell whether it contains non-ASCII characters to spot the non_English apps.

---
We built the function `is_english()` which uses the built-in `ord()` function to find out the corresponding encoding number of each character in `app_name`.

In [20]:
def is_english(app_name):
    for character in app_name:
        if ord(character) > 127:
            return False
    return True

---
Lets try out the function and see if it works as intended.

In [21]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))  # returns False
print(is_english('Instachat 😜'))  # returns False
print(ord('™'))
print(ord('😜'))

True
False
False
False
8482
128540


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.

---
To minimize the impact of data loss, we'll only remove an app (consider it non-English) if its name has more than three non-ASCII characters:

In [22]:
def is_english(app_name, num=3):
    non_ascii = 0
    
    for character in app_name:
        if ord(character) > 127:
            non_ascii += 1
            
    if non_ascii > num:
        return False
    else:
        return True

---
Lets use the modified function to check if it fits the goal.

In [23]:
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))  # now returns true
print(is_english('Instachat 😜'))  # now returns true

False
True
True


The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in our analysis — we shouldn't spend too much time on optimization at this point.


## Filtering Out Non-English Apps
---

Below, we use the `is_english()` function to filter out the non-English apps for both data sets:

In [24]:
android_english = []
ios_english = []

for row in android_clean:
    app_name = row[0]
    if is_english(app_name):
        android_english.append(row)
    
for row in ios_list:
    app_name = row[1]
    if is_english(app_name):
        ios_english.append(row)

---
Let's explore the filtered data sets stored in the lists `android_english` and `ios_english`

In [25]:
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

We can see that we're left with 9614 Android apps and 6183 iOS apps.

## Isolating the Free Apps
---
As we mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps, and we'll need to isolate only the free apps for our analysis. Below, we isolate the free apps for both our data sets:

In [26]:
android_final = []  # English apps that are free
ios_final = []  # English apps that are free

for row in android_english:
    price = row[7]
    if price == '0':
        android_final.append(row)

for row in ios_english:
    price = row[4]
    if price == '0.0':
        ios_final.append(row)

---
Let's explore our filtered data sets which should only have English apps that are free.

In [27]:
explore_data(android_final, 0, 3, True)
print('\n')
explore_data(ios_final, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 8864
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

We're left with 8864 Android apps and 3222 iOS apps, which should be enough for our analysis.

## Most Common Apps by Genre
---
As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we then develop it further.
3. If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build frequency tables for:

1. the 'Genres' and 'Category' columns of the Google Play data set
2. the 'prime_genre' column of the App Store data set
---
We'll build two functions, one to build the frequency table and another to display it.

The function `freq_table()` builds a frequency table based on the following parameters:
1. list `dataset` is the data set represented as a list
2. int `index` is the index of the column based on which we'll generate the frequency table
3. boolean `in_percent` determines whether to generate in percentage or not; default: True

We have set paramter 3 as True to generate the frequency table in percentage instead of number of occurrences.

The function `display_table()` generates a frequency table first using `freq_table()` and shows the frequency table based on the following parameters:
1. list `dataset` is the data set represented as a list
2. int `index` is the index of the column based on which we'll generate the frequency table
3. boolean `rev` determines whether to sort in descending; default: True

We have set paramter 3 as True to sort the generated frequency table in descending order of the percentage value.

In [28]:
# frequency table in percentages
def freq_table(dataset, index, in_percent=True):
    table = {}
    total = 0
    for row in dataset:
        total += 1
        value = row[index]
        table[value] = table.get(value, 0) + 1
        
    if in_percent:
        table_percentages = {}
        for key in table:
            freq = table[key]
            percentage = (freq / total) * 100
            table_percentages[key] = percentage
    
        return table_percentages
    else:
        return table
    
# display percentages in descending order
def display_table(dataset, index, rev=True):
        table = freq_table(dataset, index, True)
        table_sorted = sorted([(v, k) for k, v in table.items()], reverse=rev)
        for v, k in table_sorted:
            print(k, ':', v)

---
We start by examining the frequency table for the prime_genre column (index 11 or -5) of the "App Store" data set (stored in `ios_final`).

In [29]:
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

---
Let's continue by examining the Genres (index 9 or -4) and Category (index 1) columns of the Google Play data set (stored in `android_final`). The two columns seem to be related.

In [30]:
display_table(android_final, 1) # Category column in google play

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

---
Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column of "Google Play" data set:

In [31]:
display_table(android_final, -4) # Genres column in google play

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

## Most Popular Apps by Genre on the App Store
---

One way to find out which genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the "Installs" column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the "rating_count_tot" column.

---

Below, we calculate the average number of user ratings per app genre on the App Store:

In [32]:
ios_genres = freq_table(ios_final, -5, in_percent=False)  # prime genre
ratings_genres = {}

for genre in ios_genres:
    ratings_count = 0  # sum of number of ratings for all apps in a genre
    apps_count = 0  # total number of apps in a genre
    for row in ios_final:
        app_genre = row[-5]
        if app_genre == genre:
            num_ratings = float(row[5])
            ratings_count += num_ratings
            apps_count += 1  
    avg_num_ratings = ratings_count / apps_count
    ratings_genres[genre] = avg_num_ratings
    
# print in descending order
ratings_genres_sorted = sorted([(v, k) for k, v in ratings_genres.items()], reverse=True)
for v, k in ratings_genres_sorted:
    print(k,':' ,v)

Navigation : 86090.33333333333
Reference : 74942.11111111111
Social Networking : 71548.34905660378
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 39758.5
Food & Drink : 33333.92307692308
Finance : 31467.944444444445
Photo & Video : 28441.54375
Travel : 28243.8
Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Sports : 23008.898550724636
Games : 22788.6696905016
News : 21248.023255813954
Productivity : 21028.410714285714
Utilities : 18684.456790123455
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Business : 7491.117647058823
Education : 7003.983050847458
Catalogs : 4004.0
Medical : 612.0


On average, Navigation apps have the highest number of user reviews.

---
Let's inspect this genre.

In [33]:
for app in ios_final:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


As we can see, the figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together.

The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

---
Let's inspect the Reference genre. It has a high number of downloads and might have less number of apps from tech giants.

In [34]:
for app in ios_final:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating.

However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app ensuring more interaction.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

1. Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

2. Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

3. Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

## Most Popular Apps by Genre on Google Play
---
For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [35]:
display_table(android_final, 5) # the Installs columns

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


This data is that is not precise since we don't know if an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes.

---
We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs.

To perform computations, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [36]:
android_categories = freq_table(android_final, 1, in_percent=False) # category
installs_categories = {}

for category in android_categories:
    installs_count = 0  # sum of installs for all apps in a category
    apps_count = 0  # total number of apps in a category
    for row in android_final:
        app_category = row[1]
        if app_category == category:
            s_install = row[5]
            s_install = s_install.replace(',', '')
            s_install = s_install.replace('+', '')
            f_install = float(s_install)
            installs_count += f_install
            apps_count += 1
    avg_num_installs = installs_count / apps_count
    installs_categories[category] = avg_num_installs

# print in descending
installs_categories_sorted = sorted([(v, k) for k, v in installs_categories.items()], reverse=True)
for v, k in installs_categories_sorted:
    print(k,':' ,v)

COMMUNICATION : 38456119.167247385
VIDEO_PLAYERS : 24727872.452830188
SOCIAL : 23253652.127118643
PHOTOGRAPHY : 17840110.40229885
PRODUCTIVITY : 16787331.344927534
GAME : 15588015.603248259
TRAVEL_AND_LOCAL : 13984077.710144928
ENTERTAINMENT : 11640705.88235294
TOOLS : 10801391.298666667
NEWS_AND_MAGAZINES : 9549178.467741935
BOOKS_AND_REFERENCE : 8767811.894736841
SHOPPING : 7036877.311557789
PERSONALIZATION : 5201482.6122448975
WEATHER : 5074486.197183099
HEALTH_AND_FITNESS : 4188821.9853479853
MAPS_AND_NAVIGATION : 4056941.7741935486
FAMILY : 3695641.8198090694
SPORTS : 3638640.1428571427
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1833495.145631068
BUSINESS : 1712290.1474201474
LIFESTYLE : 1437816.2687861272
FINANCE : 1387692.475609756
HOUSE_AND_HOME : 1331540.5616438356
DATING : 854028.8303030303
COMICS : 817657.2727272727
AUTO_AND_VEHICLES : 647317.8170731707
LIBRARIES_AND_DEMO : 638503.734939759
PARENTING : 542603.6206896552
BEAUTY : 51315

Although the category "COMMUNICATION" seems very promising we should take a closer look at the apps under this category. Like the App Store there might be a handful of tech giants skewing the figures.

In [37]:
figure1 = '1,000,000,000+'
figure2 = '500,000,000+'
figure3 = '100,000,000+'
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == figure1 or app[5] == figure2 or app[5] == figure3):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs.

---

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:

In [38]:
under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3603485.3884615386

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern here is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

Like the App Store, the books and reference genre looks fairly popular here as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

---
Let's take a look at some of the apps from this genre and their number of installs:

In [39]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. 

---
It seems there's still a small number of extremely popular apps that skew the average:

In [40]:
figure1 = '1,000,000,000+'
figure2 = '500,000,000+'
figure3 = '100,000,000+'
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == figure1 or app[5] == figure2 or app[5] == figure3):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. 

---
Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [41]:
figure1 = '1,000,000+'
figure2 = '5,000,000+'
figure3 = '10,000,000+'
figure4 = '50,000,000+'
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == figure1 or app[5] == figure2 or app[5] == figure3 or app[5] == figure4):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the islamic book, the holy Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

## Conclusions

---

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.